internal/zinc-core/src/main/scala/sbt/internal/inc/Compile.scala

```diff
@@ -139,43 +144,38 @@ private final class AnalysisCallback(
       .map { case (label, map) => label + "\n\t" + map.mkString("\n\t") }
       .mkString("\n")
 
-  import collection.mutable.{ HashMap, HashSet, ListBuffer, Map, Set }
-
-  private[this] val srcs = Set[File]()
-  private[this] val classApis = new HashMap[String, (HashAPI.Hash, ClassLike)]
-  private[this] val objectApis = new HashMap[String, (HashAPI.Hash, ClassLike)]
-  private[this] val classPublicNameHashes = new HashMap[String, Array[NameHash]]
-  private[this] val objectPublicNameHashes = new HashMap[String, Array[NameHash]]
-  private[this] val usedNames = new HashMap[String, Set[UsedName]]
-  private[this] val unreporteds = new HashMap[File, ListBuffer[Problem]]
-  private[this] val reporteds = new HashMap[File, ListBuffer[Problem]]
-  private[this] val mainClasses = new HashMap[File, ListBuffer[String]]
-  private[this] val binaryDeps = new HashMap[File, Set[File]]
+  private[this] val srcs = ConcurrentHashMap.newKeySet[File]()
+  private[this] val classApis = new TrieMap[String, (HashAPI.Hash, ClassLike)]
+  private[this] val objectApis = new TrieMap[String, (HashAPI.Hash, ClassLike)]
+  private[this] val classPublicNameHashes = new TrieMap[String, Array[NameHash]]
+  private[this] val objectPublicNameHashes = new TrieMap[String, Array[NameHash]]
```

Comment:
Using an `Array` should be ok because of how the maps are updated. Elements in the array are never modified, and because we are using a concurrent map there are no visibility issues. The reason why I kept using `Array` is simply to avoid copying elements. TBH, it might be better to just use a `Vector`, as I don't think this will have a performance impact (I expect the arrays to be fairly small). Would be great to hear what the original author thinks about this.